### PR TITLE
Use Lazy<T> in DefaultFrameworkMappings

### LINKED ISSUE #8735: DefaultFrameworkNameProvider is not thread-safe
## Details about Problem

The `DefaultFrameworkNameProvider` is not thread safe so if you're creating `NuGetFramework` objects in multiple threads, its instantiation can happen multiple times.  The class should use `Lazy<T>` instead of simple `null` checks.

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): All

NuGet version (x.x.x.xxx): Any

dotnet.exe --version (if appropriate): Any

VS version (if appropriate): Any

OS version (i.e. win10 v1607 (14393.321)): Any

Worked before? If so, with which NuGet version: No

## Detailed repro steps so we can see the same problem

1. In multiple threads, parse a target framework via the `NuGetFramework` class.

...

### Verbose Logs

Please include verbose logs (NuGet.exe <COMMAND> -verbosity detailed | dotnet.exe <COMMAND> --verbosity diag | etc...)

### Sample Project

Very helpful if you can zip a project and paste into this issue!

### PR DESCRIPTION
**Pull request by jeffkl**
In the near future I'll be making performance optimizations in NuGet which include making things more parallel.  The `DefaultFrameworkNameProvider.Instance` and everything it pulls in is not thread safe which causes unnecessary overhead.

## Bug

Fixes: https://github.com/NuGet/Home/issues/8735 
Regression: No
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: Using `Lazy<T>` makes these types thread-safe and keep the delay load behavior.

## Testing/Validation

Tests Added: No  
Reason for not adding tests:  Existing tests should cover the core scenario that the class is instantiated correctly.
Validation:  I profiled NuGet while accessing this class in multiple threads and saw that it is now thread-safe.

### CODE REVIEW

**Comment by jeffkl** [comment]
Be sure to review with whitespace ignored since I didn't change any declarations, just indenting of them.

https://github.com/NuGet/NuGet.Client/pull/3089/files?utf8=%E2%9C%93&diff=split&w=1

**Review by zivkan — APPROVED** [review]
Please keep the PR template. We have a tool to generate release notes which expects it, so additional manual work will be needed if you don't use the template for the first comment of the PR, and don't have a link to an issue in NuGet/Home.

File: src/NuGet.Core/NuGet.Frameworks/DefaultFrameworkMappings.cs [hunk 1]
```diff
@@ -14,515 +14,489 @@ namespace NuGet.Frameworks
...
-                                FrameworkConstants.CommonFrameworks.NetCore451,
-                                FrameworkConstants.CommonFrameworks.DotNet53,
-                                FrameworkConstants.CommonFrameworks.NetStandard12),
-
-                            // xamarin frameworks
-                            CreateGenerationAndStandardMappingForAllVersions(
-                                FrameworkConstants.FrameworkIdentifiers.MonoAndroid,
-                                FrameworkConstants.CommonFrameworks.DotNet56,
-                                FrameworkConstants.CommonFrameworks.NetStandard21),
-
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Frameworks/DefaultFrameworkMappings.cs:466]: why remove the extra lines? :) It improves readability.

**Review by nkolev92 — APPROVED** [review]
Looks good overall. 
I have a question about the extra line removal, I think it might be unnecessary.

File: src/NuGet.Core/NuGet.Frameworks/DefaultFrameworkMappings.cs (hunk 1 above)
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Frameworks/DefaultFrameworkMappings.cs:466]: Good catch.  I didn't remove them, something must have formatted it that way.  I'll bring them back.

**Comment by jeffkl** [comment]
Can someone please start the CI jobs again?

**Comment by nkolev92** [comment]
@rrelyea 5.4 or 5.5 with these changes?

The big perf work will come in 5.5, and I'm not sure how much of a difference this would on make on it's own in 5.4. 

@jeffkl Let us know if you'd like to argue this change for 5.4.

**Review by dtivel — APPROVED** [review]

**Pull request merged**